File: 445-add-two-numbers-ii/add-two-numbers-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rev(ListNode* head){
        ListNode* temp = head;
        ListNode* prev = NULL;
        while(temp){
            ListNode* front = temp->next;
            temp->next = prev;
            prev = temp;
            temp = front;
        }
        return prev;
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* temp = rev(l1);
        ListNode* temp2 = rev(l2);
        int sum = 0, carry = 0;
        ListNode* dummy = new ListNode();
        while(temp || temp2){
            if(temp){
                sum+=temp->val;
                temp = temp->next;
            }
            if(temp2){
                sum+=temp2->val;
                temp2 = temp2->next;
            }
            dummy->val = sum%10;
            carry = sum/10;
            ListNode* node = new ListNode(carry);
            node->next = dummy;
            dummy = node;
            sum = carry;
        }
        return carry == 0 ? dummy->next : dummy;
    }
};

File: 445-add-two-numbers-ii/add-two-numbers-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        s1 = ''
        s2 = ''
        while l1:
            s1+= str(l1.val)
            l1 = l1.next
        
        while l2:
            s2+= str(l2.val)
            l2 = l2.next

        num = str(int(s1)+int(s2))
        dummy = ListNode(-1)
        temp = dummy
        for ch in num:
            temp.next = ListNode(int(ch))
            temp = temp.next
        
        return dummy.next
        